ambry-quota/src/main/java/com/github/ambry/quota/capacityunit/AmbryCUQuotaEnforcer.java

```diff
@@ -0,0 +1,141 @@
+          QuotaUtils.requestDetailsToString(restRequest, blobInfo));
+      throw new QuotaException(errorMessage, true);
+    }
+    final QuotaResource quotaResource = QuotaResource.fromRestRequest(restRequest);
+
+    doAndHandleException(() -> {
+      quotaSource.chargeUsage(quotaResource, quotaName, requestCostMap.get(quotaName));
+      return null;
+    }, String.format("Could not charge for request %s due to", restRequest));
+  }
+
+  @Override
+  public QuotaRecommendation recommend(RestRequest restRequest) throws QuotaException {
+    final QuotaName quotaName = QuotaUtils.getCUQuotaName(restRequest);
+    long usage = doAndHandleException(() -> {
+      final QuotaResource quotaResource = QuotaResource.fromRestRequest(restRequest);
+      return (long) quotaSource.getUsage(quotaResource, quotaName);
+    }, String.format("Could not recommend for request %s due to", restRequest));
+    return buildQuotaRecommendation(usage, quotaName);
+  }
+
+  @Override
+  public boolean isQuotaExceedAllowed(RestRequest restRequest) throws QuotaException {
+    return doAndHandleException(() -> quotaSource.getSystemResourceUsage(QuotaUtils.getCUQuotaName(restRequest))
+            < maxFrontendCuUsageToAllowExceed,
+        String.format("Could not check quota exceed allowed for request %s due to", restRequest));
+  }
+
+  @Override
+  public List<QuotaName> supportedQuotaNames() {
+    return SUPPORTED_QUOTA_NAMES;
+  }
+
+  @Override
+  public QuotaSource getQuotaSource() {
+    return quotaSource;
+  }
+
+  @Override
+  public void shutdown() {
+
+  }
+
+  /**
+   * Build the {@link QuotaRecommendation} object from the specified usage and {@link QuotaName}.
+   * @param usage percentage usage.
+   * @param quotaName {@link QuotaName} object.
+   * @return QuotaRecommendation object.
+   */
+  private QuotaRecommendation buildQuotaRecommendation(long usage, QuotaName quotaName) {
+    boolean shouldThrottle = (usage >= 100);
+    return new QuotaRecommendation(shouldThrottle, usage, quotaName,
+        shouldThrottle ? throttleRetryAfterMs : NO_THROTTLE_RETRY_AFTER_MS);
+  }
+
+  /**
+   * Executes the action. If the action throws an {@link Exception} this method converts it into {@link QuotaException}
+   * using the specified parameters for the error message.
+   * @param action the {@link Callable} to call.
+   * @param errorMessagePrefix exception error message to use as prefix.
+   * @param <T> the result type for {@link Callable}.
+   * @return result of the action.
+   * @throws QuotaException in case action throws an exception.
+   */
+  private <T> T doAndHandleException(Callable<T> action, String errorMessagePrefix) throws QuotaException {
+    try {
+      return action.call();
+    } catch (Exception ex) {
+      if (ex instanceof QuotaException) {
+        throw new QuotaException(String.format("%s %s", errorMessagePrefix, ex.getMessage()),
+            ((QuotaException) ex).isRetryable());
+      }
```

Comment:
this seems unnecessary.  we can just throw this QuotaException.

Reply:
wait, I am talking about this if statement. when the exception is an instance of QuotaException, we don't have to convert it again to QuotaException, we can just throw it out. unless what you are trying to write in if condition is `!(ex instanceof QuotaException)`

Reply:
I want to add more context to the exception's message. At some point (most likely in `Chargeable` implementation), we will log the QuotaException message as error logs when needed. I want that log to contain enough information to that it will be easy for us to understand what happened through the chain of calls.